src/transformStateWithClones.js

```diff
@@ -7,7 +8,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+
+  let newObject = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    if (type === 'addProperties') {
+      newObject = { ...newObject, ...extraData };
+    }
+
+    if (type === 'removeProperties') {
+      for (const item of keysToRemove) {
+        delete newObject[item];
+      }
+    }
+
+    if (type === 'clear') {
+      for (const key in newObject) {
+        delete newObject[key];
+      }
+    }
+
+    result.push({ ...newObject });
```

Comment:
Pushing a copy of the object to the result array should be done once after the 'switch' block, not within each case. This will adhere to the DRY (Don't Repeat Yourself) principle.